### ISSUE TITLE
hypre is called in a c function, which is called in Fortran main program, result: each processor create and set its own matrix, vector and solver

### ISSUE THREAD

**Issue by suguanting** [hypre-space/hypre#174]
Hi,

I wish to use HYPRE within a main porject written in FORTRAN. And I am having trouble calling HYPRE function using FORTRAN code (some unknown trouble to get FORTRAN resolve the HYPRE functions), therefore I try to circumvent the problem by first set up the HYPRE solver within a C funtion (named PPE_HYPRE_MPI) and then call that C function in the FORTRAN main program. The pseudo code look something like this:
MAIN.f90:
`PROGRAM MAIN`
`...`
`CALL PPE_HYPRE_MPI (...)`
`...`
`STOP`
`END`

PPE_HYPRE_MPI.cpp:
`#include "selfmade_hypre_header.h"`
`extern "C" void ppe_hypre_mpi(...)`
`{`
`HYPRE_StructGrid     grid;`
`HYPRE_StructStencil  stencil;`
`HYPRE_StructMatrix   A;`
`HYPRE_StructVector   b;`
`HYPRE_StructVector   x;`
`HYPRE_StructSolver   solver;`
`HYPRE_StructSolver   precond;`
`...`
`}`

selfmade_hypre_header.h:
`#include <stdio.h>`
`#include <stdlib.h>`
`#include <string.h>`
`#include <math.h>`
`#include "HYPRE_struct_ls.h"`

This mixed language project works perfectly for a single processor, however, it cannot achieve the multiprocessing as expected. For example, when using two processor, they are actually solving two different problems, each with half of the problem defintion (below is the matrix A coeffcients processor 0 and processor 1 output after the matrix A is assembled using HYPRE_StructMatrixAssemble, note that  different halves of the matrix is zero for different processor)
![image](https://user-images.githubusercontent.com/14161266/89668571-8dcdcb80-d910-11ea-8e67-3f855610d58c.png)
![image](https://user-images.githubusercontent.com/14161266/89669033-5875ad80-d911-11ea-97b5-a49f71cf107f.png)

I notice that API states that `grid` should be defined on "global abstract index space". I guess this just mean they should be treated as global variables, so I moved those declarations to the header file "selfmade_hypre_header.h", but it doesn't fix the problem. And I further noticed that in the examples, objects like `grid` or `stencil` are declared before MPI is initiated. However, this is quite impossible as the main program is written in FORTRAN, where MPI is initiated. I am not sure that one can interoperate variable like `grid` between the two languages. 

Could anyone please tell me where I got things wrong and is there a solution to this? 

To some extent, I am also aware that the root cause of such a problem probably is that I try to use HYPRE in a external function. Actually, I would love to get rid of this feature if I could use HYPRE in FORTRAN. I am using Visual studio and Intel compilers for C and Fortran. As you can tell, I can run HYPRE in C, but failed to do so for FORTRAN through the same approach. Nominally, "HYPRE.lib" is linked, but FORTRAN won't resolve any HYPRE funtion. I have looked up ex12f.f and ex5f.f and add `include 'HYPREf.h'`, but it is still not working. I could not find a FORTRAN couterpart for "HYPRE_struct_ls.h" from which FORTRAN could know about the functions included in "HYPRE.lib". I would be very glad if someone could tell me how to get FORTRAN resolve HYPRE or which part of the manual I should referred to. 

Thank you so much. I really appreciate your help,

Guanting

**Comment by rfalgout** [comment]
Hi @suguanting .  Example `ex12f.f` should be a good example to look at for what you want to do.  The issue with the Fortran compiler not seeing the hypre function calls is a simple name mangling issue that we can help you with.  Can you send us the error messages you get when you try to link `ex12f.f`?  Note that `SStruct` and `Struct` are very similar when the grid is a simple structured grid; the difference is mainly the extra `S` and one or two additional arguments in the function calls (`part` and `var`).

**Comment by suguanting** [comment]
Hi Rob, thank you for the reply, the name mangling issue is solved and I am now using hypre in Fortran. However, I have encountered a similar problem and I have found out that the real problem is that hypre is producing different results when different numbers of processor are used. This problem occur when Neumann type boundary conditions are used. I will close this issue and open up a new issue to concentrate on the real problem.

Thank again,
Guanting
